### PR TITLE
feat: Add Environment & Services support

### LINKED ISSUE #212: update for e&s
**Is your feature request related to a problem? Please describe.**

Environments & Services are coming to Honeycomb. The beeline should be updated to support sending data to environments while maintaining backward compatibility. To achieve this, we will set dataset from service name, and provide appropriate warnings depending on environment vs. non-environment teams.

**Describe the solution you'd like**

Tracing scenario for environment-aware teams:

* Dataset should no longer propagate across services for new environment key.
* `ServiceName` "required"†. If unset, will warn and set `service_name` AND `service.name` to `unknown_service:<process>` or `unknown_service:language`. Service name is used to populate dataset name; if `unknown_service.*` then truncate to`unknown_service` for dataset name. `service.name` field will be added to the already added `service_name` for closer consistency with OTel.
* `WriteKey` required. If unset, will warn.
* `Dataset` **ignored**. If unset, no warning (because no longer used). If set, will warn to clarify that it will be ignored in favor of service name.

Tracing scenario for legacy (non-environment-aware) teams:

* `ServiceName` "required"†. If unset, will warn.
* `WriteKey` required. If unset, will warn.
* `Dataset` required. If unset, will warn and default to `beeline-<language>`.

† While not technically required, service name is highly encouraged to avoid `unknown_service` and to instead properly describe the data being sent to Honeycomb (and the data being viewed in Honeycomb UI).

This will be addressed in a separate issue/PR: Set [spanKind value based on OTel spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind) for consistency across beelines and OTel.

### Notes from decision around defaulting when no service name is provided:
*Given* a beeline configured with a v2 api key and without a service name
*When* it sends HNY events
*Then* it should use `unknown_service` as the dataset
*And* it should use `unknown_service:process-name` as service_name

When given a legacy key, the behavior remains the same as today.

[Notes on trimming whitespace](https://honeycomb.quip.com/qdrvA6a9AHEI/notes-trimming-whitespace-ingest)

### PR DESCRIPTION
**Pull request by JamieDanielson**
## Which problem is this PR solving?

- closes #212 

## Short description of the changes

Tracing scenario for environment-aware teams:

* Dataset no longer propagates across services for new environment key.
* Add field `service.name` in addition to already added `service_name`
* `ServiceName` "required"†. If unset, will warn and set `service_name` AND `service.name` to `unknown_service:<process>` or `unknown_service:language`.
* Service name is used to populate dataset name; if `unknown_service.*` then truncate to`unknown_service` for dataset name.
* Trim whitespace from dataset name
* `WriteKey` required. If unset, will warn.
* `Dataset` **ignored**. If unset, no warning (because no longer used). If set, will warn to clarify that it will be ignored in favor of service name.

Tracing scenario for classic (non-environment-aware) teams:

* Add field `service.name` in addition to already added `service_name`
* `ServiceName` "required"†. If unset, will warn.
* `WriteKey` required. If unset, will warn.
* `Dataset` required. If unset, will warn and default to `beeline-<language>`.

† While not technically required, service name is highly encouraged to avoid `unknown_service` and to instead properly describe the data being sent to Honeycomb (and the data being viewed in Honeycomb UI).

logging when missing environment variables:

```
ERROR:root:writekey not set! set the writekey if you want to send data to honeycomb
ERROR:root:service name not set! service name will be unknown_service:python
ERROR:root:data will be sent to unknown_service
```

### CODE REVIEW

File: beeline/__init__.py [hunk 1]
```diff
@@ -74,15 +75,55 @@ def __init__(self,
...
+                self.log(
+                    'dataset not set! set a value for dataset if you want to send data to honeycomb')
+        else:
+            if dataset:
+                self.log(
+                    'dataset will be ignored in favor of service name'
+                )
 
+        default_service_name = "unknown_service"
+        process_name = os.environ.get('PROCESS_EXECUTABLE_NAME', '')
```
Comment (JamieDanielson) [inline beeline/__init__.py:101]: [upstream otel-python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/resources/__init__.py#L173) only checks for env var of process name, so mirrored that here

File: beeline/propagation/test_honeycomb.py [hunk 2]
```diff
@@ -21,6 +22,7 @@ def test_roundtrip(self):
 
     def test_roundtrip_with_dataset(self):
         '''Verify that we can successfully roundtrip (marshal and unmarshal)'''
+        beeline.propagation.propagate_dataset = True
```
Comment (vreynolds) [inline beeline/propagation/test_honeycomb.py:25]: ask: can we add tests for disabled propagation

**Review by vreynolds — CHANGES_REQUESTED** [review]

File: beeline/__init__.py [hunk 3]
```diff
@@ -74,15 +75,55 @@ def __init__(self,
...
             self._init_logger()
 
+        def IsClassicKey(writekey):
+            return len(writekey) == 32
+
         # allow setting some values from the environment
         if not writekey:
             writekey = os.environ.get('HONEYCOMB_WRITEKEY', '')
+        if not writekey:
+            self.log(
```
[the remark this answers is not in the document]
Comment (JamieDanielson) [reply beeline/__init__.py:85]: The way the logging is set up in this beeline currently, it [only logs these messages](https://github.com/honeycombio/beeline-python/blob/main/beeline/__init__.py#L74) if `debug=True`.  Actually now wondering if we should change the behavior to match the changes in other beelines that always warn on missing config (and send to stderr).

File: beeline/__init__.py [hunk 4]
```diff
@@ -74,15 +75,55 @@ def __init__(self,
...
+                'service name not set! service name will be ' + service_name
+            )
+            if not IsClassicKey(writekey):
+                self.log(
+                    'data will be sent to unknown_service'
+                )
+
+        if not IsClassicKey(writekey):
+            # overwrite dataset with service name, trim whitespace
+            dataset = service_name.strip()
```
[the remark this answers is not in the document]
Comment (JamieDanielson) [reply beeline/__init__.py]: https://github.com/honeycombio/beeline-python/pull/213/commits/66f13db59577fd442140ba2cc7ab4f144a48f68e

File: beeline/propagation/test_honeycomb.py (hunk 2 above)
Comment (JamieDanielson) [reply beeline/propagation/test_honeycomb.py:25]: https://github.com/honeycombio/beeline-python/pull/213/commits/6f876be7ec211e43af08da3b1f042642ba9b72c6

File: examples/hello-world/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+[tool.poetry]
+name = "hello-world"
+version = "0.1.0"
+description = "print hello world"
+authors = ["honeycombio"]
+
+[tool.poetry.dependencies]
+python = "^3.9"
+honeycomb-beeline = {path = "../../dist/honeycomb_beeline-3.42.99-py3-none-any.whl", develop = false}
```
Comment (vreynolds) [inline examples/hello-world/pyproject.toml:9]: 🤩 ooh, is `whl` the magic suffix to get local reference to work?

**Review by vreynolds — APPROVED** [review]
slytherin

File: examples/hello-world/pyproject.toml (hunk 5 above)
Comment (JamieDanielson) [reply examples/hello-world/pyproject.toml:9]: No 😭  I had hoped it would but it works the same as a regular tar.gz file.

**Pull request merged**